blog/2023-10-25-implementing-sign-in-with-google.md

```diff
@@ -0,0 +1,519 @@
+    "in `camelCase` as well",
+)
+```
+### GoogleLogin
+
+The `GoogleLogin` component renders the familiar "Sign in with Google" button.
+
+Since we will use the default configuration, no props are needed, however the
+event trigger does need to be defined so our Reflex app is able to get the token
+after logging in.
+
+Define the following wrapper in the same `react_oauth_google.py`.
+
+```python
+class GoogleLogin(rx.Component):
+    library = "@react-oauth/google"
+    tag = "GoogleLogin"
+
+    def get_event_triggers(self):
+        return \{"on_success": lambda data: [data]}
+```
+
+The `on_success` trigger is defined to pass its argument directly to the Reflex
+event handler.
+
+## Handling the Token
+
+```python eval
+docalert(
+    "Always use TLS in production",
+    "The user's token may be compromised if sent over an insecure websocket!",
+    status="warning",
+)
+```
+
+An event handler will be used to receive the token after a successful login.
+Critically, the token must be verified and decoded to access the user
+information it contains.
+
+### google-auth
+
+The simplest way to verify the token is to use Google's own `google-auth` python
+library. Add `google-auth[requests]` to your app's `requirements.txt` and install it with
+`pip install -r requirements.txt`.
+
+Add the necessary imports to the module where your app `State` is defined and
+set the `CLIENT_ID` saved earlier, as it is needed to verify the token.
+
+```python
+from google.auth.transport import requests
+from google.oauth2.id_token import verify_oauth2_token
+
+CLIENT_ID = "309209880368-4uqd9e44h7t4alhhdqn48pvvr63cc5j5.apps.googleusercontent.com"
+```
+
+### on_success
+
+The `on_success` trigger is fired by `GoogleLogin` after a successful login, and it contains the `id_token` that provides user information. For now, this event handler will verify the token and dump its contents to the console to verify that it is working.
+
+```python
+class State(rx.State):
+    ...
+    def on_success(self, id_token: dict):
+        print(
+            verify_oauth2_token(
+                id_token["credential"],
+                requests.Request(),
+                CLIENT_ID
+            )
+        )
+```
+
+## Logging In
+
+With this minimal functionality in place, it should be possible to log in with
+one of the test accounts defined earlier on the Consent Screen configuration.
+
+Add the `GoogleOAuthProvider` and `GoogleLogin` components linked with the
+previously defined `CLIENT_ID` and `State.on_success` event handler to test the
+functionality so far.
+
+```python
+from .react_oauth_google import GoogleOAuthProvider, GoogleLogin
+
+def index():
+    return rx.vstack(
+        GoogleOAuthProvider.create(
+            GoogleLogin.create(on_success=State.on_success),
+            client_id=CLIENT_ID
+        )
+    )
+
+app = rx.App()
+app.add_page(index)
+app.compile()
+```
+
+After a successful login, you will see the decoded JWT with user profile
```

Comment:
Might not be obvious what JWT is for a beginner

Reply:
added a link to jwt.io and expanded the acronym inline